lib/validation.js

```diff
@@ -0,0 +1,86 @@
+const vscode = require('vscode');
+const { validateHubl } = require('@hubspot/cli-lib/api/validate');
+const { getPortalId } = require('@hubspot/cli-lib');
+const {
+  TEMPLATE_ERRORS_TYPES,
+  VSCODE_SEVERITY,
+  WORD_REGEX,
+} = require('./constants');
+const fs = require('fs');
+const path = require('path');
+
+const getRange = (document, error) => {
+  const adjustedLineNumber = error.lineno > 0 ? error.lineno - 1 : 0;
+
+  if (error.startPosition > 0) {
+    return document.getWordRangeAtPosition(
+      new vscode.Position(adjustedLineNumber, error.startPosition - 1),
+      WORD_REGEX
+    );
+  } else {
+    return document.lineAt(new vscode.Position(adjustedLineNumber, 0)).range;
+  }
+};
+
+const isFileInWorkspace = (error) => {
+  const pathToActiveFile = vscode.window.activeTextEditor.document.uri.path;
+  const dirToActiveFile = path.dirname(pathToActiveFile);
+
+  let filePath = error.categoryErrors.fullName || error.categoryErrors.path;
+
+  if (error.category === 'MODULE_NOT_FOUND') {
+    filePath = filePath + '.module';
+  }
+
+  return fs.existsSync(path.resolve(dirToActiveFile, filePath));
+};
+
+const clearValidation = (document, collection) => {
+  collection.set(document.uri, null);
+};
+
+const updateValidation = async (document, collection) => {
+  if (!document) {
+    return collection.clear();
+  }
+
+  const { renderingErrors } = await validateHubl(
+    getPortalId(),
+    document.getText()
+  );
+
+  const fitleredErrors = renderingErrors.filter((error) => {
+    if (
+      error.reason === TEMPLATE_ERRORS_TYPES.MISSING ||
+      error.reason === TEMPLATE_ERRORS_TYPES.BAD_URL
+    ) {
+      return isFileInWorkspace(error) ? false : true;
+    }
+    return true;
+  });
+
+  const templateErrors = fitleredErrors.map((error) => {
+    return {
+      code: '',
+      message: error.message,
+      range: getRange(document, error),
+      severity: vscode.DiagnosticSeverity[VSCODE_SEVERITY[error.reason]],
+    };
+  });
+
+  collection.set(document.uri, templateErrors);
+};
+
+let timeout = null;
+const triggerValidate = (document, collection) => {
+  if (!['html-hubl', 'css-hubl'].includes(document.languageId)) {
+    return clearValidation(document, collection);
```

Comment:
Do we also want to call `clearTimeout(timeout)` here? I'm not sure if it's possible, but I'm imagining a scenario where [this collection.set](https://github.com/HubSpot/hubspot-cms-vscode/pull/95/files#diff-419296896634fcf982f1a147262973ede2ee05a726db022c357eafa6c6ed7ecfR71) gets called after we call `clearValidation` here b/c of the 1s timeout OR possibly b/c the `validateHubl` call could take long.

I haven't played around with this too much yet, so this race condition might not actually exist.